pkg/task/task.go

```diff
@@ -43,200 +46,274 @@ import (
-// newTaskExecutor returns a new instance of taskExecutor
-func newTaskExecutor(runtask *v1alpha1.RunTask, values map[string]interface{}) (*taskExecutor, error) {
-	mte, err := newMetaTaskExecutor(runtask.Spec.Meta, values)
+// newExecutor returns a new instance of
+// executor
+func newExecutor(rt *v1alpha1.RunTask, values map[string]interface{}) (*executor, error) {
+	mte, err := newMetaExecutor(rt.Spec.Meta, values)
 	if err != nil {
-		return nil, err
+		return nil,
+			errors.Wrapf(err, "failed to init task executor: failed to init meta executor: values {%+v}: %s", values, rt)
 	}
 
-	return &taskExecutor{
-		templateValues: values,
-		metaTaskExec:   mte,
-		runtask:        runtask,
+	return &executor{
+		Values:   values,
+		MetaExec: mte,
+		Runtask:  rt,
 	}, nil
 }
 
-// String provides the essential task executor details
-func (m *taskExecutor) String() string {
-	return fmt.Sprintf("task with identity '%s' and with objectname '%s'", m.getTaskIdentity(), m.getTaskObjectName())
+// String is the Stringer implementation
+// of executor
+func (m *executor) String() string {
+	return stringer.Yaml("task executor", m)
+}
+
+// GoString is the GoStringer implementation
+// of executor
+func (m *executor) GoString() string {
+	return stringer.Yaml("task executor", m)
 }
 
 // getTaskIdentity gets the task identity
-func (m *taskExecutor) getTaskIdentity() string {
-	return m.metaTaskExec.getIdentity()
+func (m *executor) getTaskIdentity() string {
+	return m.MetaExec.getIdentity()
 }
 
 // getTaskObjectName gets the task's object name
-func (m *taskExecutor) getTaskObjectName() string {
-	return m.metaTaskExec.getObjectName()
+func (m *executor) getTaskObjectName() string {
+	return m.MetaExec.getObjectName()
 }
 
-// getTaskRunNamespace gets the task's run namespace
-func (m *taskExecutor) getTaskRunNamespace() string {
-	return m.metaTaskExec.getRunNamespace()
+// getTaskRunNamespace gets the namespace where
+// RunTask should get executed
+func (m *executor) getTaskRunNamespace() string {
+	return m.MetaExec.getRunNamespace()
 }
 
 // getK8sClient gets the kubernetes client to execute this task
-func (m *taskExecutor) getK8sClient() *m_k8s_client.K8sClient {
-	return m.metaTaskExec.getK8sClient()
+func (m *executor) getK8sClient() *m_k8s_client.K8sClient {
+	return m.MetaExec.getK8sClient()
 }
 
-// Output returns the result of templating this task's yaml
+// Output returns the result of templating a
+// RunTask meant for templating only purpose
 //
-// This implements TaskOutputExecutor interface
-func (m *taskExecutor) Output() (output []byte, err error) {
-	output, err = template.AsTemplatedBytes("Output", m.runtask.Spec.Task, m.templateValues)
-	return
+// NOTE:
+//  This implements OutputExecutor interface
+func (m *executor) Output() ([]byte, error) {
+	output, err := template.AsTemplatedBytes(
+		"output",
+		m.Runtask.Spec.Task,
+		m.Values,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to generate output: %s", m)
```

Comment:
Unit test of `Output`, will be helpful in verifying the errors.

Reply:
also there are pointers in the fields of executer, make sure they are not nil either at caller side, or here.